fuel-merkle/src/sparse/verify.rs

```diff
@@ -0,0 +1,341 @@
+use crate::{
+    common::{
+        path::{
+            Instruction,
+            Path,
+        },
+        Bytes32,
+        Prefix,
+    },
+    sparse::{
+        empty_sum,
+        proof::{
+            ExclusionProof,
+            InclusionProof,
+            Proof,
+        },
+        Node,
+    },
+};
+
+pub fn verify<K: Into<Bytes32>, V: AsRef<[u8]>>(key: K, value: &V, proof: Proof) -> bool {
+    match proof {
+        Proof::InclusionProof(proof) => verify_inclusion(key, value, proof),
+        Proof::ExclusionProof(proof) => {
+            value.as_ref() == empty_sum() && verify_exclusion(key, proof)
+        }
+    }
+}
+
+fn verify_inclusion<K: Into<Bytes32>, V: AsRef<[u8]>>(
+    key: K,
+    value: &V,
+    proof: InclusionProof,
+) -> bool {
+    let InclusionProof { root, proof_set } = proof;
+
+    let key: Bytes32 = key.into();
+    let leaf = Node::create_leaf(&key, value);
+    let path = leaf.leaf_key();
+    let mut current = *leaf.hash();
+
+    for (i, side_hash) in proof_set.iter().enumerate() {
+        let index = u32::try_from(proof_set.len() - 1 - i).expect("Index is valid");
+        let prefix = Prefix::Node;
+        current = match path.get_instruction(index).unwrap() {
+            Instruction::Left => Node::calculate_hash(&prefix, &current, side_hash),
+            Instruction::Right => Node::calculate_hash(&prefix, side_hash, &current),
+        };
+    }
+
+    current == root
+}
+
+fn verify_exclusion<K: Into<Bytes32>>(key: K, proof: ExclusionProof) -> bool {
+    let ExclusionProof {
+        root,
+        proof_set,
+        path,
+        hash,
+    } = proof;
+
+    if key.into() != path {
+        return false;
+    }
```

Comment:
`key` already is a path, we don't need to store it inside of the `ExclusionProof`

Reply:
The point of storing the path (which is originally a key) in the `ExclusionProof` is to make sure we verifying exclusion for the same key used to generate the exclusion proof. Because we can generate exclusion proofs for placeholders, proof generation may generate the same proof for multiple keys with placeholder values. Therefore, we may want a way to differentiate the proofs based on a unique property which, in this case, is the key.